src/Engine/Combat/BattleInstance.java

```diff
@@ -59,42 +62,85 @@ public BattleInstance(Unit pAttacker, Unit pDefender, GameMap map, int attackerX
    */
   public BattleSummary calculateBattleResults()
   {
+    CombatContext context = new CombatContext(gameMap, attacker, attackerWeapon, defender, defenderWeapon, battleRange, attackerX, attackerY);
+    Map<Unit, Entry<Weapon,Double>> unitDamageMap = new HashMap<Unit, Entry<Weapon,Double>>();
+    unitDamageMap.put(attacker, new AbstractMap.SimpleEntry<Weapon,Double>(attackerWeapon, 0.0));
```

Comment:
This whole `AbstractMap.SimpleEntry<>` thing works, but is not idiomatic or fun to read - I might suggest making a custom class to encapsulate the relationship you are representing. You are basically abusing `Map` to make a genericized tuple class - from what I've heard, the designers of Java specifically didn't include a generic Pair/Tuple class because it's semantically meaningless; better to define your own relationships with more meaning.
You also don't need a `Map` here at all - you will always be associating two sets of three things.

Reply:
That is, until we add units with splash damage or something... cross that bridge when (IF) we get there.

Reply:
Hm, that's fair. I can poke at it a bit more, I guess.

Reply:
>You are basically abusing `Map` to make a genericized tuple class - from what I've heard, the designers of Java specifically didn't include a generic Pair/Tuple class because it's semantically meaningless; better to define your own relationships with more meaning.

I prefer using generic tuples than defining a class to be used in exactly one piece of code, but maybe that's just me.
The duplication between BattleInstance and CombatContext is because one of them should be immutable, the other should mutable. I could make CombatContext a child of BattleInstance with getters and setters, but I didn't feel like rewriting a bunch of code :P

>You also don't need a `Map` here at all - you will always be associating two sets of three things.

Hmm. Okay. I took another look at the code and I'm prepared to argue my case. I'm not against changing it, necessarily, but hear me out here.
....I *will* agree that this could be better documented whatever we decide, though.
I don't need arbitrary capacities; that's entirely true. I *do* need an "unhinged" mapping for the values, though, and a collection is the easiest way I know of (least LoC and conditionals) to do that.
I want to update the value associated with the attacker or the defender, but at compile time I can't know whose damage was just calculated. All I can do is match Unit instances and call it a day.

> That is, until we add units with splash damage or something... cross that bridge when (IF) we get there.

I recommend against giving me ideas like this. I might just go implement them, instead of adding more core functionality.

Reply:
Define "unhinged" - the attacker/defender don't change after this point, so there shouldn't be a problem with using two objects for `attackerDamageStats` and `defenderDamageStats` or whatever, unless I'm missing something. Why can't you know whose damage was just calculated at compile-time?

I ain't gonna make you change it, since it's only used in this small section of code, and it's easy enough to figure it out. I'm just saying if I ever have a reason to touch this code, I might not be able to resist changing it myself.

And speaking of lines of code - counting just straight lines in files (non-syntax-aware counting) we seem to be solidly over 16k now. So that's cool.

Reply:
> Define "unhinged" - the attacker/defender don't change after this point, so there shouldn't be a problem with using two objects for `attackerDamageStats` and `defenderDamageStats` or whatever, unless I'm missing something. Why can't you know whose damage was just calculated at compile-time?

"Unhinged" in that I can just put in or pull out a value corresponding to a particular object, and have "someone else" manage which object it is.
You're missing something (unless I am?). Venge *can* swap the attacker and defender just a few lines down, at line 72. I no longer know that the attacker and defender in the `CombatContext` are the same as those in my `BattleInstance` variables.
The `Params` give me the damage for "attacker" and "defender", but I don't know which one is which anymore. So I just plop the battle results in a map and call it a day.

> I ain't gonna make you change it, since it's only used in this small section of code, and it's easy enough to figure it out. I'm just saying if I ever have a reason to touch this code, I might not be able to resist changing it myself.

Make a class to be used in one function? I guess if you're touching it again, it would be more likely to be used outside of that...

> And speaking of lines of code - counting just straight lines in files (non-syntax-aware counting) we seem to be solidly over 16k now. So that's cool.

Hah, swanky. ~~Does that count the AWBW branch?~~

Reply:
> Venge can swap the attacker and defender just a few lines down, at line 72

The `unitDamageMap` is not read for a while, it's written again first on line 86, which is after the calls to `changeCombatContext`. The creation of `unitDamageMap` could be moved to there and work just the same. Though I think it is possible with the current architecture, I don't think you intend for `applyCombatModifiers` to be used for swapping combatants.

But I'll concede there isn't much need to change it in this case. In general though, idiomatic code is nice.

Reply:
> Though I think it is possible with the current architecture, I don't think you intend for `applyCombatModifiers` to be used for swapping combatants.

It occurs to me that I could restrict this by making BattleInstance the combatRef, but ugh...

Reply:
Yeah, I feel like this whole deal could still benefit from some restructuring, but it's working for now at least.